File: go/linkedlist/leetcode/add_two_numbers_ii.go
package main

// 两数相加II

// 反转之后，就是两数相加的问题
func addTwoNumbersII(l1 *ListNode, l2 *ListNode) *ListNode {
	p := reverse(l1)
	q := reverse(l2)
	res := addTwoNumbers(p, q)
	return reverse(res)
}

func reverse(head *ListNode) *ListNode {
	var prev *ListNode
	for head != nil {
		head.Next, head, prev = prev, head.Next, head
	}
	return prev
}

// 反序处理，可以利用栈
func addTwoNumbersII2(l1 *ListNode, l2 *ListNode) *ListNode {
	var p []*ListNode
	for l1 != nil {
		p = append(p, l1)
		l1 = l1.Next
	}

	var q []*ListNode
	for l2 != nil {
		q = append(q, l2)
		l2 = l2.Next
	}

	var res *ListNode
	count := 0
	for len(p) > 0 || len(q) > 0 {
		if len(p) > 0 {
			node := p[len(p)-1]
			p = p[:len(p)-1]
			count += node.Val
		}
		if len(q) > 0 {
			node := q[len(q)-1]
			q = q[:len(q)-1]
			count += node.Val
		}
		node := &ListNode{Val: count % 10}
		node.Next = res
		res, count = node, count/10
	}
	if count > 0 {
		node := &ListNode{Val: count}
		node.Next = res
		res = node
	}
	return res
}

File: cpp/linkedlist/add_two_numbers_ii.cpp
//
// Created by roseduan on 2021/9/18.
// 两数相加II
//
#include "vector"
#include "stack"
using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution{
public:
    // 使用栈
    ListNode* addTwoNumbersII(ListNode* l1, ListNode* l2) {
        stack<int> s1, s2;
        while (l1) {
            s1.push(l1->val);
            l1 = l1->next;
        }
        while (l2) {
            s2.push(l2->val);
            l2 = l2->next;
        }

        ListNode *p = nullptr;
        int count = 0;
        while (!s1.empty() || !s2.empty()) {
            if (!s1.empty()) {
                count += s1.top();
                s1.pop();
            }
            if (!s2.empty()) {
                count += s2.top();
                s2.pop();
            }
            auto node = new ListNode(count % 10);
            node->next = p;
            p = node;
            count /= 10;
        }
        if (count > 0) {
            auto node = new ListNode(count);
            node->next = p;
            p = node;
        }
        return p;
    }
};

int main() {
    auto *node1 = new ListNode(100);
    auto *node2 = new ListNode(100);
    Solution sol;
    sol.addTwoNumbersII(node1, node2);
}
